File: 11.go
package main

import "fmt"

var max int = 0

func calc(numbers [400]int, index func(int, int) (int, int, bool)) {
    var current int
    var count int
    for i := 0; i < 20; i++ {
        // filter 0
        current = 1
        count = 0
        for j := 0; j < 20; j++ {
            now, first, err := index(i, j)
            if err {
                break
            }
            if numbers[now] == 0 {
                current = 1
                count = 0
                continue
            }
            current *= numbers[now]
            if count < 4 {
                count ++
            } else {
                current /= numbers[first]
            }
            if current > max && count == 4 {
                max = current
            }
        }
    }
}

func by_row(i, j int) (int, int, bool) {
    return i * 20 + j, i * 20 + j - 4, false
}

func by_col(i, j int) (int, int, bool) {
    return j * 20 + i, (j - 4) * 20 + i, false
}

func by_diagonal_left_top(i, j int) (int, int, bool) {
    if i < 4 || i < j {
        return 0, 0, true
    }
    return (i - j) * 20 + j, (i - (j - 4)) * 20 + (j - 4), false
}

func by_diagonal_right_bottom(i, j int) (int, int, bool) {
    if i == 0 || i > 16 || 19 - i < j {
        return 0, 0, true
    }
    return (19 - j) * 20 + i + j, (19 - (j - 4)) * 20 + i + j - 4, false
}

func by_diagonal_right_top(i, j int) (int, int, bool) {
    if i < 4 || i < j {
        return 0, 0, true
    }
    return j * 20 + (19 - i) + j, (j - 4) * 20 + (19 - i) + j - 4, false
}

func by_diagonal_left_bottom(i, j int) (int, int, bool) {
    if i == 0 || i > 16 || 19 - i < j {
        return 0, 0, true
    }
    return (i + j) * 20 + j, (i + j - 4) * 20 + j - 4, false
}

func main() {
    var numbers [400]int
    // read numbers
    for i := 0; i < 400; i++ {
        fmt.Scanf("%d", &numbers[i])
    }
    // calculate by lines
    calc(numbers, by_row)
    // calculate by columns
    calc(numbers, by_col)
    // calculate by diagonal lines
    calc(numbers, by_diagonal_left_top)
    calc(numbers, by_diagonal_right_bottom)
    calc(numbers, by_diagonal_right_top)
    calc(numbers, by_diagonal_left_bottom)
    fmt.Printf("%d\n", max)
}

File: 11.c
#include <stdio.h>

int max = 0;

typedef struct {
    unsigned int current;
    unsigned int first;
    int error:1;
} IndexValue;

void calc(int *numbers, IndexValue(*index)(int, int)) {
    int current, count;
    IndexValue index_value;
    for (int i = 0; i < 20; i++) {
        current = 1;
        count = 0;
        for (int j = 0; j < 20; j++) {
            index_value = index(i, j);
            if (index_value.error) {
                break;
            }
            if (numbers[index_value.current] == 0) {
                current = 1;
                count = 0;
                continue;
            }
            current *= numbers[index_value.current];
            if (count < 4) {
                count++;
            } else {
                current /= numbers[index_value.first];
            }
            if (current > max && count == 4) {
                max = current;
            }
        }
    }
    return;
}

IndexValue by_row(int i, int j) {
    return (IndexValue){i * 20 + j, i * 20 + j - 4, 0};
}

IndexValue by_col(int i, int j) {
    return (IndexValue){j * 20 + i, (j - 4) * 20 + i, 0};
}

IndexValue by_diagonal_left_top(int i, int j) {
    if (i < 4 || i < j) {
        return (IndexValue){0, 0, 1};
    }
    return (IndexValue){(i - j) * 20 + j, (i - (j - 4)) * 20 + (j - 4), 0};
}

IndexValue by_diagonal_right_bottom(int i, int j) {
    if (i == 0 || i > 16 || 19 - i < j) {
        return (IndexValue){0, 0, 1};
    }
    return (IndexValue){(19 - j) * 20 + i + j, (19 - (j - 4)) * 20 + i + j - 4, 0};
}

IndexValue by_diagonal_right_top(int i, int j) {
    if (i < 4 || i < j) {
        return (IndexValue){0, 0, 1};
    }
    return (IndexValue){j * 20 + (19 - i) + j, (j - 4) * 20 + (19 - i) + j - 4 , 0};
}

IndexValue by_diagonal_left_bottom(int i, int j) {
    if (i == 0 || i > 16 || 19 - i < j) {
        return (IndexValue){0, 0, 1};
    }
    return (IndexValue){(i + j) * 20 + j, (i + j - 4) * 20 + j - 4, 0};
}

int main(int argc, char *argv[]) {
    int numbers[20][20];
    for (int i = 0; i < 20; i++) {
        for (int j = 0; j < 20; j++)
        {
            scanf("%d", &numbers[i][j]);
        }
    }
    calc(numbers, by_row);
    calc(numbers, by_col);
    calc(numbers, by_diagonal_left_top);
    calc(numbers, by_diagonal_right_bottom);
    calc(numbers, by_diagonal_right_top);
    calc(numbers, by_diagonal_left_bottom);
    printf("%d\n", max);
    return 0;
}